Introduce the provider property

mattcaswell opened the pull request:
Replace the properties default, fips and legacy with a single property
called "provider". So, for example, instead of writing "default=yes" to
get algorithms from the default provider you would instead write
"provider=default". We also have a new "fips" property to indicate that
an algorithm is compatible with FIPS mode. This applies to all the
algorithms in the FIPS provider, as well as any non-cryptographic
algorithms (currently only serializers).

slontis commented on providers/defltprov.c:
```diff
@@ -331,134 +331,138 @@ static OSSL_ALGORITHM exported_ciphers[OSSL_NELEM(deflt_ciphers)];
+ */
 static const OSSL_ALGORITHM deflt_serializer[] = {
-    { "RSA", "default=yes,format=text,type=private",
+    { "RSA", "provider=default,fips=yes,format=text,type=private",
```
I am confused..
When we do a query with fips=n o doesnt that mean it wont find the serializer??

levitte replied:
Er, I concur.  Why should the **default** provider have _anything_ with the property `fips=yes`?

mattcaswell replied:
Because, "fips=yes" now means that the algorithm is compatible for use in FIPS mode. The serializers in the default provider *are* compatible and hence have the property. A query of "fips=no" means only algorithms that are not compatible for use in FIPS mode, and would rightly exclude the serializers. If you instead mean only those algorithms that are not in the FIPS *provider* then the query string should be "provider!=fips".

paulidale replied:
I cannot think of a case where _FIPS=no_ is used.  Perhaps non-US governments might but they'll not be asking for a NIST certified algorithm and thus they won't want a FIPS serialiser.

levitte replied:
The absence of the fips property doesn't mean "fips=no", does it? I gotta admit that I'm a bit confused on that...

richsalz replied:
I think knowing whether or not an algorithm is FIPS does not belong in OpenSSL; it belongs in the FIPS documents -- are you promising to keep current whenever they issue an update, for example? If it *does* end up in OpenSSL, I think it should be a separate stand-alone function, table-driven and global. It should not be part of the property lookup mechanism.

beldmit replied:
Sorry, I don't get a point. 

We have some algorithms that are allowed according to FIPS documents. 
We have various implementations of them. Some of them are FIPS-certified.

These 2 are different properties. What does "fips=yes" mean?

mattcaswell replied:
"fips=yes" means it is allowed to be used in FIPS mode. This is different to it being an algorithm that is implemented in the FIPS provider (although it includes all of those). Essentially in FIPS mode you can use all the algorithms in the FIPS provider plus the serializers.

richsalz replied:
>"fips=yes" means it is allowed to be used in FIPS mode.

NO.  This is wrong.  What is allowed to be used in FIPS mode is *FIPS-validated implementations.*  Algorithms that are acceptable are enumerated by NIST, but you can't use a non-validated implementation in FIPS mode.  This will cause immense confusion.

slontis replied:
> I think knowing whether or not an algorithm is FIPS does not belong in OpenSSL; it belongs in the FIPS documents -- are you promising to keep current whenever they issue an update, for example?

You are entitled to your own view.
A fips provider is a list of certified algorithms..
e.g: https://csrc.nist.gov/projects/cryptographic-module-validation-program/Certificate/2473
It doesnt need to implement every single algorithm there is at any point in time - it can just be one algorithm if that is all that is needed.
Fips modules can also contain non fips algorithms (that can be accessed in non fips mode) if you really want.

You will also find the FIPS_mode() and FIPS_mode_set() listed in the design document.

richsalz replied:
> You will also find the FIPS_mode() and FIPS_mode_set() listed in the design document.

And a wrapper mode for engines.  Your point being?

levitte replied:
Regarding engines, we chose to leave them as they are.  A wrapper proved to be much harder than anticipated, and time runeth away.  That doesn't mean that we exclude a _future_ wrapper, though.

As far as I'm personally concerned, I wouldn't include `FIPS_mode()` or `FIPS_mode_set()` at all, but it was concluded very early (in the first design meeting) that there those that will want to keep using these.

So from an application point of view, ENGINEs are still going to be there in 3.0 as before, and `FIPS_mode()` / `FIPS_mode_set()` as well.

mattcaswell replied:
> NO. This is wrong. What is allowed to be used in FIPS mode is FIPS-validated implementations. Algorithms that are acceptable are enumerated by NIST, but you can't use a non-validated implementation in FIPS mode. This will cause immense confusion.

You do realise that what we are talking about here are the serializers? If you disable access to the serializers then you won't be able to:

- Save a generated private key for later use
- Send a public key to anyone
- Construct certificates
- Use TLS/CMS or any other similar protocol that transmits public keys between peers

FIPS without these capabilities is useless. AFAIK the OpenSSL FOM 2.0 did not disable access to these things.

levitte replied:
There's a lot of side discussion in this comment thread.  My mistake picking up one of those side tracks...
The original concern seems to be resolved as far as I can tell, so...

paulidale commented:
I need to consider this far more deeply.

mspncp commented:
Let’s discuss this on Tuesday.

mattcaswell commented:
It seems like lots of people are getting very confused about what we're trying to do here. Let me restate the problem:

PR #11075, implements the function `FIPS_mode_set()`. It does this by setting the default property query "fips=yes". This means that any algorithm implementation that does not have this property against it will not be fetched. However this causes a problem because it means the serializers (which are not in the FIPS provider) will not be fetched - and yet it is perfectly ok for an application in fips mode to use those serializers. The problem is that the set of algorithms that are ok to use in FIPS mode is  not the same as the set of algorithms that are in the FIPS provider.  The set is bigger than just that.

Part of the problem is that our existing "fips" property means both "the set of algorithms loaded from the FIPS provider" *and* "the set of algorithms you can use in FIPS mode". This PR splits those two concepts.

levitte commented:
There's a problem with setting the property "fips=yes" in the default provider, though... It requires that the default provider gets loaded in the first place, and that's something we've explicitly avoided the need for whith the fips provider. This PR is not up to the point in its current form.

What's needed is a separate serializer provider. I would prefer to see it built in, for convenience. Do note that we want the same serializers available through the default provider too, for performance reasons. In the end, all that will differ with regards to the serializers will be which algorithm table we go through and the properties used in each.

mattcaswell commented:
> There's a problem with setting the property "fips=yes" in the default provider, though... It requires that the default provider gets loaded in the first place, and that's something we've explicitly avoided the need for whith the fips provider. This PR is not up to the point in its current form.

This PR does not attempt to solve the complete problem. It does solve *part* of the problem.

> What's needed is a separate serializer provider. I would prefer to see it built in, for convenience. Do note that we want the same serializers available through the default provider too...

All of this I fully agree with, and had that end goal in mind when writing this PR.

It should be possible to run in FIPS mode with either:
1) The FIPS and default provider loaded
2) The FIPS and separate serializer provider loaded
3) Just the FIPS provider and no serializer support

This PR allows (1) and (3), and will be needed for (2) to work.

levitte commented:
I may have overstated that a bit.  What I'm trying to say is that "fips=yes" in the default provider is incorrect, at least if we look at the objective that the default provider doesn't hold fips-usable algorithms, period.

levitte commented:
The consequence is that if you want serializers when in "fips mode", you will *have to* load the separate serializer module.  Of course.

mattcaswell commented:
>  at least if we look at the objective that the default provider doesn't hold fips-usable algorithms

Since when has that been an objective? Our objective has been to have a FIPS provider that can be validated. There is nothing in that which restricts what can go in the default provider.

> The consequence is that if you want serializers when in "fips mode", you will have to load the separate serializer module.

I don't agree with your premise, and therefore obviously I don't agree with your consequence.

levitte commented:
So if we're going to have "fips=yes" on the serializers that are in the default provider, then I don't see why we need a separate serializer provider, at all.  All that will do is duplicate almost exactly the same algorithm table for them, which seems quite unnecessary under these circumstances.

Note: in this "fips mode" vs "non-fips mode" debate, I don't see any importance in the provider name.  Yeah, they can still be nice to have if you do want to select algorithms based on that, and I do think that this PR fulfills that objective perfectly.

mattcaswell commented:
> So if we're going to have "fips=yes" on the serializers that are in the default provider, then I don't see why we need a separate serializer provider, at all.

Yes, I've considered this point. However I think there will be some people that will not want the default provider loaded at all - to avoid the possibility of "accidents". There will be others that may want the default provider loaded so that they can perform both FIPS operations and non-FIPS operations in the same application. So I think it is desirable to have both.

> All that will do is duplicate almost exactly the same algorithm table for them, which seems quite unnecessary under these circumstances.

Since they would both be "built-in" providers I think we could arrange things so that they both share the same physical table for the serializers - no duplication required.

> Note: in this "fips mode" vs "non-fips mode" debate, I don't see any importance in the provider name. Yeah, they can still be nice to have if you do want to select algorithms based on that, and I do think that this PR fulfills that objective perfectly.

Exactly.

levitte commented:
> Since they would both be "built-in" providers I think we could arrange things so that they both share the same physical table for the serializers - no duplication required.

You're right! The only real difference would be the first downcall table, the one given back by the knit function.... and obviously and init function itself.

richsalz commented:
Based on https://github.com/openssl/openssl/pull/11097#issuecomment-586637695, PR #11075 is a mistake and should not be merged.

mattcaswell commented:
> Based on #11097 (comment), PR #11075 is a mistake and should not be merged.

#11075 is not the problem - it just exposed it. The problem exists anyway. This PR fixes the immediate issue. With this PR in place #11075 is perfectly fine to merge.

richsalz commented:
My main concern is at https://github.com/openssl/openssl/pull/11097#discussion_r379837254

slontis commented:
> Based on [#11097 (comment)](https://github.com/openssl/openssl/pull/11097#issuecomment-586637695), PR #11075 is a mistake and should not be merged.

The FIPS_mode_set in this PR calls an existing public API that can set properties per libctx (so yes the problem already existed).

richsalz commented:
> Part of the problem is that our existing "fips" property means both "the set of algorithms loaded from the FIPS provider" and "the set of algorithms you can use in FIPS mode". This PR splits those two concepts.

I view "fips" as meaning "this provider has been validated."  If my application cares about that the provider implements, it will either know out-of-band by reading the documentation (such as the OpenSSL architecture docs), or it will query to make sure the provider does the algorithms it wants.

It is a mistake to think of "FIPS mode" as something that still exists. What exists is providers, and which one(s) the application loads and uses.  That is also why I think "add FIPS_mode API" is wrong.  We have been saying that we don't *enforce* things in code, and FIPS-mode leads one to think that this is what is happening. Moving the "don't use non-FIPS implementations" is being left as administrative guidance.

richsalz commented:
> I may have overstated that a bit. What I'm trying to say is that "fips=yes" in the default provider is incorrect, at least if we look at the objective that the default provider doesn't hold fips-usable algorithms, period.

And I believe that this will lead to serious confusion and widespread mistakes amongst OpenSSL users who want FIPS-validated cryptography.

richsalz commented:
> The FIPS_mode_set in this PR calls an existing public API that can set properties per libctx (so yes the problem already existed).

Yes, I get it, but it's still wrong to have that API with its *extensive* baggage present in this release. Since FOREVER, that function locked you into FIPS mode; it was a one-way switch, it couldn't be reversed. That's not what we're doing in this release, and there's no need for this outdated API, that is a convenience wrapper around an existing API, to exist with different semantics than what the entire OpenSSL FIPS-user community is used to.

paulidale commented:
After more thought, I'm okay with this approach so I'm unblocking.

Another possibility would be to allow "fips=*" in a property definition.  The "*" being a wildcard that matches regardless of what is asked for.  I.e. it would match both "fips=yes" and "fips=no" queries.  I don't think I'd have it add to the score of an optional "?fips" however.  I don't think it is as streamlined as the solution here.

paulidale commented on crypto/property/property_parse.c:
```diff
@@ -579,7 +576,11 @@ int ossl_property_parse_init(OPENSSL_CTX *ctx)
-        || (ossl_property_false = ossl_property_value(ctx, "no", 1)) == 0)
+        || (ossl_property_false = ossl_property_value(ctx, "no", 1)) == 0
+        /* Create values for the built-in providers */
+        || ossl_property_value(ctx, "default", 1) == 0
```
These aren't required.  Values are created on demand when parsing definitions.

The reason that _yes_ and _no_ are treated specially is to avoid lots of look ups when checking Booleans.  With this change we're moving away from Booleans and we could reconsider having these special values entirely.

mattcaswell replied:
I backed this change out.

paulidale commented on doc/man7/property.pod:
```diff
@@ -118,7 +118,7 @@ Ordering of optional clauses is not significant.
 In order to permit a more concise expression of boolean properties, there
 is one short cut: a property name alone (e.g. "default") is
-exactly equivalent to "default=yes" in both definitions and queries.
+exactly equivalent to "provider=default" in both definitions and queries.
```
I'm quite uncomfortable with this short cut.  Booleans by default seems okay, providers a lot less so.

It is also not implemented :)

mattcaswell replied:
This is actually more a case of me not properly reading the text I was updating rather than an intention to implement this short cut!

mattcaswell replied:
Fixed.

mattcaswell commented:
Fixup pushed addressing the latest feedback.

levitte commented on crypto/property/property_parse.c:
```diff
@@ -562,12 +562,9 @@ OSSL_PROPERTY_LIST *ossl_property_merge(const OSSL_PROPERTY_LIST *a,
         "version",      /* Version number of this provider */
-        "fips",         /* FIPS supporting provider */
-        "engine",       /* An old style engine masquerading as a provider */
+        "fips",         /* FIPS compatible algorithm */
```
Nit: I like "supporting" better than "compatible"

mspncp replied:
Or „FIPS conforming“?

levitte replied:
Not sure...  do the serializers conform to FIPS in any kind of way?

mattcaswell replied:
I like "supporting"

mattcaswell replied:
Fixed this, and tweaked the wording in a few other spots too. Fixup pushed.

levitte commented:
I've warmed up to this PR.  Looks fine to me

levitte commented:
I would like to see another approval beside mine.  I know that it's more than what our criteria require, but safety in numbers, or something like that...

richsalz commented:
What does the "fips" property indicate?  Suppose there are two FIPS-validated providers that don't implement all the same algorithms?

mattcaswell commented:
> What does the "fips" property indicate?

An algorithm that is allowed for use in FIPS mode (either because it is a validated algorithm in itself, or is an allowed supporting algorithm). In practice it means a validated algorithm or a serializer.

The documentation in this PR that describes how "fips=yes" is used against algorithms in the default provider as follows:

"Some non-cryptographic algorithms (such as serializers for loading keys and
parameters from files) are not FIPS algorithm implementations in themselves but
support algorithms from the FIPS provider and are allowed for use in "FIPS
mode". The property "fips=yes" can be used to select such algorithms."

> Suppose there are two FIPS-validated providers that don't implement all the same algorithms?

I'm not too sure of the relevance of this.

richsalz commented on doc/man7/property.pod:
```diff
@@ -52,15 +52,15 @@ The full syntax for property definitions appears below.
 for all of its algorithms.
-Likewise, the FIPS provider defines I<fips=yes> and the legacy provider
-defines I<legacy=yes> for all of their algorithms.
+Likewise, the FIPS provider defines I<provider=fips> and the legacy provider
```
Is there only one FIPS provider?  Should that be "a FIPS provider"?  And didn't @slontis say that a FIPS-validated provider could have non-FIPS algorithms in it?

levitte replied:
It should probably say "OpenSSL's FIPS provider"

paulidale replied:
I'm okay leaving it as is or using @levitte's suggestion.

The FIPS provider can and I suspect will have non-validated algorithms, it is still the FIPS provider.

mattcaswell replied:
I fixed this using @levitte's suuggestion

richsalz commented on providers/defltprov.c:
```diff
@@ -20,7 +20,7 @@
 #include "internal/nelem.h"
 
-#define ALGC(NAMES, FUNC, CHECK) { { NAMES, "default=yes", FUNC }, CHECK }
+#define ALGC(NAMES, FUNC, CHECK) { { NAMES, "provider=default", FUNC }, CHECK }
```
Is this syntactic sugar/shorthand really necessary?

levitte replied:
Yes. I have an upcoming script that needs something like this.

richsalz commented:
Labelling a non-crytographic algorithm as "FIPS" is going to be confusing.  Labelling anything other than a FIPS-possible algorithm as "FIPS" is a bad idea.

You folks are too involved in this to see the big picture.

FIPS means validated implementation of a NIST-approved algorithm.  Do not redefine the term.  Create a new one if it's necessary.

t-j-h commented:
Actually "FIPS" means "The Federal Information Processing Standard" and there are a large number of those documents. It does not mean FIPS140-2 validated (or the prior FIPS140-1 or the pending FIPS140-3).  

Our use here is rather loose ... and this isn't aimed at your comment - just noting that use of "fips140" would be better ...

slontis commented:
> Actually "FIPS" means "Federal Information Processing Standard" and there are a large number of those documents.

Agreed..  The term 'fips module' has been used by openssl for quite a while, which implies this is the case.

paulidale commented:
With the potential opening of a major can of worms: might we consider provider dependencies?  Then the _default_ and _FIPS_ providers could both depend on _serialisers_.

I think this would be well out of scope for 3.0.

paulidale commented on doc/man7/provider.pod:
```diff
@@ -260,25 +260,31 @@ algorithm identifier to the appropriate fetching function.
-be used as a search criterion for these implementations.
+implementations of the same algorithms), the property "provider=fips" can
+be used as a search criterion for these implementations. All algorithm
+implementations in the FIPS provider can also be selected with the property
```
I'm still expecting e.g. MD5 to be in the FIPS provider (for TLS).  It can't be selected using "fips=yes".  It could be selected using "provider=fips" and then we're back to the beginning of this mess.

I don't think that this should prevent this PR moving forward.

richsalz replied:
> I'm still expecting e.g. MD5 to be in the FIPS provider (for TLS)

I don't see how.  Looking at the IGhttps://csrc.nist.gov/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf page 15, it says that if SHA1 is available than MD5 must not be used, for example.

paulidale replied:
Several TLS versions require both SHA1 and MD5 to be present.

slontis replied:
These fall into the category of "Non-FIPS Approved But Allowed Cryptographic Functions" See SP 800-135

t8m replied:
Is TLS-1.0, 1.1 still something to worry about for FIPS? I would say that with the dependence of these protocol versions on SHA1 and prevalence of TLS-1.2 they do not have to be supported in FIPS mode anymore.

mattcaswell commented:
> Do not redefine the term. Create a new one if it's necessary.

I'm open to considering alternative name suggestions (which could be done either in this PR or a follow up one)

richsalz commented:
It's not related to FIPS cryptographic algorithms, so a name like "fips-unrelated" or "not-cryptographic" might make sense.

mattcaswell commented:
> It's not related to FIPS cryptographic algorithms, so a name like "fips-unrelated" or "not-cryptographic" might make sense.

Its intended to convey the union of the algorithms in the (a?) FIPS module along with those other serializer type algorithms. So I'm not sure those names quite cover it.

richsalz commented:
I don't see a serializer as an algorithm.  FIPS means crypto.  Perhaps something along the lines of auxiliary or assistive works.

mattcaswell commented:
> I don't see a serializer as an algorithm. 

In OpenSSL terms they are. Providers provide algorithms communicated through an array of `OSSL_ALGORITHM` entries. There is one such `OSSL_ALGORITHM` entry for each serializer.

But whether you think of them as an algorithm or not doesn't really matter. The name we use should cover the union of algorithms in the FIPS module along with the serializers (whatever they are).

levitte commented:
So essentially, we make "fips=yes" mean "whatever is needed to enable FIPS validated algorithms, without introducing non-FIPS algorithms", i.e. it's not a strict definition like  what @richsalz tries to imply.

This comes down to practicality, and I can't see that as bad thing.

richsalz commented:
I don't need serializers to use a FIPS validated algorithms or provider.

You are taking a word which has a common meaning and extending the definition. I do not see why you think that's not a problem, why you do not think it will cause confusion for OpenSSL's users.

How about "facility" or "subsystem"?

A serializer is not an algorithm.

mattcaswell commented:
> You are taking a word which has a common meaning and extending the definition.

I'm not sure which word we're arguing about here "fips" or "algorithm".

If it's "algorithm" then that word is used extensively throughout IT. The Cambridge dictionary defines it as:

"a set of mathematical instructions or rules that, especially if given to a computer, will help to calculate an answer to a problem"

https://dictionary.cambridge.org/dictionary/english/algorithm

In this case the problem that serializers solve is "given an internal representation of a key/parameter give me back that key/parameter in well known format X". Our providers implement a set of instructions to solve that problem. They are very clearly algorithms in the common use of the term. They are probably not "cryptographic algorithms" in that they don't solve a cryptographic problem such as "key exchange", "signature", "encryption" etc.

levitte commented:
> I don't need serializers to use a FIPS validated algorithms or provider.

That's what you think...  it happens under the hood, any time there's a part of libcrypto that needs it for a key that only exists in the provider.  For example, a call of `i2d_PrivateKey` will call a provided serializer when the passed `EVP_PKEY` doesn't have an assigned legacy key, just an assigned provided key.

levitte commented:
`i2d_PrivateKey` was a bad example, as private keys are supposed not to escape from the FIPS provider.  However, there are serializers of public keys and parameters too, to text (used by functions like `EVP_PKEY_print_public()` and `EVP_PKEY_print_params()`, which use the text serializers, `X509_PUBKEY_set()`, which uses the public key to DER serializers, ...

So er, you might think that you don't need serializers, but...

mattcaswell commented:
> i2d_PrivateKey was a bad example, as private keys are supposed not to escape from the FIPS

Not so...its perfectly ok for private keys generated by the FIPS provider to be serialized into a file.

levitte commented:
> Not so...its perfectly ok for private keys generated by the FIPS provider to be serialized into a file.

Oh, right, then can be encrypted (PKCS#8 organization is assumed)

richsalz commented:
They *can* be used and perhaps they are used by the implementation in some corner cases, but my application does not need to use serializers.

As for terminology, I am saying that "FIPS algorithm" should not include seriaizers, but only mean crypto algorithms, and using the "fips" selector for this stuff is wrong.

mattcaswell commented:
> They can be used and perhaps they are used by the implementation in some corner cases, but my application does not need to use serializers.

Are you doing TLS? If so, you need them.

richsalz commented:
> Are you doing TLS? If so, you need them.

in my code?  Or does OpenSSL use them?

levitte commented:
> in my code? Or does OpenSSL use them?

libcrypto needs them, and by extension if not directly, so does libssl.  Does your application use those libraries?  Then the need extends to your application as well.

levitte commented:
You *could* of course make a study of exactly which of all possible code paths may or may not need some kind of serializations.  Chances are, though, that as soon as you deal with certificates and their public keys, you're getting there, because `X509_PUBKEY_set()`

richsalz commented:
I am sorry I am not explaining this well enough to make my point clear.

*MY* code does not use serializers.  *YOUR CODE* in its implementation of TLS does.  So what.  Why should the application care?  Any more than why should the application care about BIGNUM or ```bn->top```?  *it should not.*

We are, of course, arguging about pedantic side-points.  That's what programmers do.

"FIPS algorithm" means "crypto" and *only crypto.*  Saying otherwise is misleading to everyone not initimately familiar with the details of the project.  You are doing your users a dis-service here.

mattcaswell commented:
> in my code? Or does OpenSSL use them?

libssl will use the serializers. For example in a KeyExchange message or ClientHello key_share, libssl will ask the ephemeral DH/ECDH public key to serialize itself to send to the peer (not all of this code is there yet...but it will be by the time of 3.0).

levitte commented:
So @richsalz, if you dislike "fips=yes" so much for the purpose we (i.e. libcrypto and libssl) need, what do you suggest instead that's indicative enough?

(trying very hard to get out of the trenches)

richsalz commented:
I already did:
https://github.com/openssl/openssl/pull/11097#issuecomment-587025545
https://github.com/openssl/openssl/pull/11097#issuecomment-587006862

levitte commented:
@paulidale said:
> With the potential opening of a major can of worms: might we consider provider dependencies?

Actually, @mattcaswell and I were discussing an idea along those lines, but instead of introducing dependencies, which may indeed be a can of worms, we were thinking of a possible autoload feature, on a per operation basis if I recall correctly.  So for example, if nothing could be found for the operation `OSSL_OP_SERIALIZER` (i.e. no currently loaded provider would answer to that query), we would have programming indicating that the code should then try to load the "serializer" provider and try once more.

We even discussed making it possible to have the autoload feature configurable.

levitte commented:
> I already did:
> https://github.com/openssl/openssl/pull/11097#issuecomment-587025545
> https://github.com/openssl/openssl/pull/11097#issuecomment-587006862

Thanks.  Not sure what values would be expected for "assistive", "auxilliary", "facility" or "subsystem".  None of them are particularly intuitive...  But then, what about something far more generic?  "tag=*something*"?  "tag=fips"?  "tag=fips-support"?

richsalz commented:
Autoload?  Doesn't this repeat the discussion of autoloading providers which we decided not to do by default?

richsalz commented:
> What about ...

Anything other than "fips"  The values would be which type of provider as in "assists=fips,default" for example.

t8m commented:
I have to say I am perfectly fine with fips=yes being indicative of something (crypto or non-crypto algorithm) that when used from OpenSSL is FIPS compliant. I do not know why the definition must be "fips approved crypto algorithm" and why any other definition would be confusing. And I am really scared by the @richsalz alternative proposals.

levitte commented:
> Autoload? Doesn't this repeat the discussion of autoloading providers which we decided not to do by default?

Kinda sorta...  we decided not to use built-in fallback providers when there are already loaded providers.  So this autoload feature would actually work in a similar manner, but on the more fine grained on-operation basis instead of on-whole-provider basis.

levitte commented:
@richsalz said:
> The values would be which type of provider as in "assists=fips,default" for example.

So a property query would have to be something that covers ( "fips=yes" xor "assist=fips" ) to satisfy you.  Trouble, we don't have the necessary property query syntax for it...  the closest we have would be "?fips=yes,?assist=yes", and I'm actually unsure what happens if none if the components match.  @paulidale?

mattcaswell commented:
> I have to say I am perfectly fine with fips=yes being indicative of something (crypto or non-crypto algorithm) that when used from OpenSSL is FIPS compliant. I do not know why the definition must be "fips approved crypto algorithm" and why any other definition would be confusing. And I am really scared by the @richsalz alternative proposals.

I could easily see a number of properties being defined, e.g.

fips_alg = a validated FIPS algorithm
fips_supporting = a non-FIPS algorithm that is allowed to be used in a FIPS compliant application
fips = The union of "fips_alg" and "fips_supporting"

I can't currently think of a better name for the union property other than just "fips". The suggestions so far don't seem right to me.

t8m commented:
> I could easily see a number of properties being defined, e.g.
> 
> fips_alg = a validated FIPS algorithm
> fips_supporting = a non-FIPS algorithm that is allowed to be used in a FIPS compliant application
> fips = The union of "fips_alg" and "fips_supporting"

I could certainly live with that although the usefulness of the two extra properties does not seem obvious to me.

levitte commented:
> the usefulness of the two extra properties does not seem obvious to me.

To me neither, but they are worth talking about to get out of the "you're wrong!  Am not!  Are too!" type of trench...

mattcaswell commented:
I pushed a fixup to address the minor doc nit noted above. Please reconfirm!

richsalz commented:
The trench is my digging uphill to convince the project that using "fips" for this is wrong.  I lost.  That's okay, I lose most of my discussion points.  Still doesn't make me wrong nor you right. :)

paulidale commented:
> the closest we have would be "?fips=yes,?assist=yes"

This will match neither.  It prefers having the properties but not having them isn't a failure.

levitte commented:
That's what I suspected

t8m commented:
Still approved.

mattcaswell commented:
I was holding this one back to ensure #10631 could go in without merge conflicts. Now that has been merged, I've rebased and fixed the conflict issues.

@levitte / @t8m / @paulidale please can you reconfirm?

paulidale commented on doc/man7/property.pod:
```diff
@@ -117,8 +117,8 @@ Ordering of optional clauses is not significant.
 In order to permit a more concise expression of boolean properties, there
-is one short cut: a property name alone (e.g. "default") is
-exactly equivalent to "default=yes" in both definitions and queries.
+is one short cut: a property name alone (e.g. "myproperty") is
```
A thought: should _myproperty_ be changed to something that is legal.  E.g. _my.property_.  User defined properties need a dot in them.

This is completely optional.

paulidale commented on providers/defltprov.c:
```diff
@@ -331,136 +331,140 @@ static OSSL_ALGORITHM exported_ciphers[OSSL_NELEM(deflt_ciphers)];
 
+/*
+ * Unlike most algorithms in the default provider, the serializers are allowed
+ * for use in FIPS mode, and therefore have the "fips=yes" property.
```
Is it worth mentioning something along the lines of: _because they are not FIPS relevant_ here?

mattcaswell commented:
I addressed both issues raised by @paulidale.

levitte commented on test/recipes/30-test_evp_fetch_prov.t:
```diff
@@ -60,28 +60,28 @@ unless ($no_fips) {
+              { args      => [ '-property', 'provider!=fips' ],
+                message   => 'using property "provider!=fips"' },
+              { args      => [ '-property', 'provider=fips' ],
+                message   => 'using property "provider=fips"' } ] }
```
Actually, these tests aren't really very interesting. Don't get me wrong, of course we should test the boring properties, but the actually interesting one to test is "fips" with all its possible values, not to mention "-fips". With this change, that property, which be more crucial than any "provider=", is not tested at all.

In my opinion, this is subject for dismissal of approvals made so far.

mattcaswell replied:
I've added  a bunch of new test cases to this test to cover the "fips" property.

mattcaswell commented:
@levitte - fixup pushed. Please take another look.

levitte commented on doc/man7/EVP_MAC-CMAC.pod:
```diff
@@ -15,7 +15,7 @@ used with EVP_MAC_fetch():
 =over 4
 
-=item "CMAC", "default=yes"
+=item "CMAC", "provider=default"
```
Also "provider=fips"

levitte commented on doc/man7/EVP_MAC-GMAC.pod:
```diff
@@ -15,7 +15,7 @@ used with EVP_MAC_fetch():
 =over 4
 
-=item "GMAC", "default=yes"
+=item "GMAC", "provider=default"
```
Also "provider=fips"

levitte commented on doc/man7/EVP_MAC-HMAC.pod:
```diff
@@ -15,7 +15,7 @@ used with EVP_MAC_fetch():
 =over 4
 
-=item "HMAC", "default=yes"
+=item "HMAC", "provider=default"
```
Also "provider=fips"

levitte commented on doc/man7/EVP_MAC-KMAC.pod:
```diff
@@ -16,9 +16,9 @@ properties, to be used with EVP_MAC_fetch():
 =over 4
 
-=item "KMAC-128", "default=yes"
+=item "KMAC-128", "provider=default"
```
Also "provider=fips"

levitte commented on doc/man7/EVP_MAC-KMAC.pod:
```diff
@@ -16,9 +16,9 @@ properties, to be used with EVP_MAC_fetch():
+=item "KMAC-128", "provider=default"
 
-=item "KMAC-256", "default=yes"
+=item "KMAC-256", "provider=default"
```
Also "provider=fips"

levitte commented:
That was the last nits from my end, I think

mattcaswell commented:
All nits fixed and fixup pushed.

levitte commented:
@paulidale, @t8m, you might want to re-approve

mattcaswell commented:
Travis red cross is not relevant.

mattcaswell commented:
Pushed. Thanks to everyone involved!

richsalz commented:
Sorry to write on a closed PR, does this require an update to the "EVP Configuration Module" section of doc/man5/config.pod ?

levitte commented:
I don't think so.

richsalz commented:
Hopefully someone who really understands what this PR did, can comment more definitively. :)

levitte commented:
Er... Not sure how to interpret what you just wrote.

The point is that "fips=yes" still means the same thing, at least mechanically. The philosophical discussion in "the meaning of FIPS" has done nothing to change that.

richsalz commented:
I didn't like "I don't think so."  If you're convinced no update is needed, that's fine.

levitte commented:
Is my expanded answer good enough?

richsalz commented:
> Is my expanded answer good enough?

Yes, it is definitive.
